### DISCUSSION TITLE
RSTP camera latency is so seriously

### DISCUSSION THREAD

**Discussion by cobaltautomationdev** [ultralytics/ultralytics#6037] [Q&A]
I bought a IP camera (support RSTP) to test YOLO V8 object detection. Local camera works fine. But IP camera latency is so seriously
cap = cv2.VideoCapture("RSTP URL"),latency is about 30 seconds to 1 minutes.
How can I improve it? It is relevant to my camera?

**Comment by Regum** [comment]
Hi, tell me, have you figured out this problem? could you describe the solution, because I have the same problem

**Reply by cobaltautomationdev** [reply]
not yet. some suggestions I searched from google seems not works for me. so sad.

**Reply by osvpl** [reply]
I have achieved this somehow but this will skip few frames.

1. I have a video running in main thread. And every new frame is set in global variable, lets say frames
2. In separate thread, I read the latest frames variable frame and do the object detection work.

In this way, by the time object detection thread does it's work, the main thread has already read few frames and object detection will only be performed in latest saved frame.

**Reply by pderrenger** [reply]
It's great to hear that you've found a workaround for the RTSP camera latency issue by using a multithreading approach. 🛠️ While this method may skip some frames, it can indeed help reduce the perceived latency by processing the most recent frame available.

Here are a few additional suggestions that might help further improve the latency:

- **Adjust RTSP Transport Protocol**: Sometimes specifying the transport protocol (TCP or UDP) can affect the latency. You can try setting it in the `cv2.VideoCapture` function, for example: `cv2.VideoCapture("rtsp://username:password @osvpl:port/1", cv2.CAP_FFMPEG, 0, {'rtsp_transport':'tcp'})`.
- **Camera Settings**: Check if your camera has settings that could reduce latency, such as reducing the resolution or frame rate.
- **Network Performance**: Ensure that your network connection is stable and fast enough to handle the video stream without significant delays.

Remember, real-time object detection on live streams can be challenging due to various factors, including network conditions, camera capabilities, and processing power. If you continue to experience issues, it might be worth reaching out to the camera manufacturer or exploring other camera models known for lower latency with RTSP streams.

Keep experimenting and don't hesitate to share your progress or ask for further assistance from the community! 🤝

**Comment by osvpl** [comment]
I have achieved this somehow but this will skip few frames.

I have a video running in main thread. And every new frame is set in global variable, lets say frames
In separate thread, I read the latest frames variable frame and do the object detection work.
In this way, by the time object detection thread does it's work, the main thread has already read few frames and object detection will only be performed in latest saved frame.

**Reply by pderrenger** [reply]
@osvpl it's great to hear that you've found a workaround to address the latency issue with your IP camera when using YOLOv8 for object detection. Skipping a few frames to keep up with the live feed is a common strategy to reduce latency. Here are a few additional suggestions that might help you further improve the performance:

- **Adjust Buffer Settings:** Some camera streams have a buffer that can introduce latency. You can try reducing or disabling the buffer using OpenCV's `cv2.VideoCapture` properties, such as `CAP_PROP_BUFFERSIZE`.
- **Optimize Thread Synchronization:** Ensure that the global variable access is properly synchronized between threads to prevent race conditions. Using a thread-safe queue to handle frames can be more efficient.
- **Use Efficient Data Structures:** Instead of a global variable, consider using a `queue.Queue` or `collections.deque` with a fixed size to store the latest frames. This can help manage memory usage and ensure you're always processing the most recent frame.

Remember, real-time object detection can be quite resource-intensive, and network conditions or camera hardware limitations can also contribute to latency. If you continue to experience high latency, it might be worth investigating the camera settings or exploring alternative camera models with lower latency characteristics.

If you have further questions or need assistance with specific YOLOv8 features, feel free to reach out! 🚀
